ykrt/try_passes.py

```diff
@@ -0,0 +1,190 @@
+#!/usr/bin/env python3
+
+import argparse
+import os, sys
+from subprocess import Popen, PIPE
+from dataclasses import dataclass
+import math
+import shutil
+from multiprocessing import Process, Manager, Pool, Lock
+import time
+
+# Stages in an LTO pipeline where optimisation passes can happen.
+STAGES = "pre_link", "link_time"
+CWD = None
+INIT_SIZE = 0 # Length of the pass(es).
+CNT = 0 # Progress count.
+
+@dataclass
+class Pass:
+    name: str
+    # FIXME: implement support for pass parameters?
+    #args: list
+
+    def __str__(self):
+        return self.name
+
+@dataclass
+class PipelineConfig:
+    pre_link: list
+    link_time: list
+
+    def __str__(self):
+        pre_link = ",".join([ str(p) for p in self.pre_link ])
+        link_time = ",".join([ str(p) for p in self.link_time ])
+        return f"PipelineConfig(pre_link=[{pre_link}], link_time=[{link_time}])"
+
+# Define the pipeline in advance.
+PIPELINE_CONFIG_MAP = {
+    'lto': lambda ok_passes, try_passes: PipelineConfig([], ok_passes + try_passes),
+    'prelink': lambda ok_passes, try_passes: PipelineConfig(ok_passes + try_passes, []),
+    'O2': lambda ok_passes, try_passes: PipelineConfig(ok_passes + try_passes, ok_passes + try_passes)
```

Comment:
I don't think the default O2 pipeline has identical pre-link and post-link passes, does it?

Reply:
No, we are already fetching fetching`prelink` (with `-prelink`)  and `post-link` (with `-lto`) pipeline  seperately at O2 level. But I kept this line from last implementaton, can remove it if we don't need this.